src/lib/Serialize.hs

```diff
@@ -133,3 +217,81 @@ cached cacheName key create = do
       BS.writeFile cacheKeyPath $ encode (curCompilerVersion, key)
       BS.writeFile cachePath    $ encode result
       return result
+
+-- === instances ===
+
+tp :: (HasPtrs a, Applicative f) => (PtrType -> RawPtr -> f RawPtr) -> a -> f a
+tp = traversePtrs
+
+instance HasPtrs Expr where
+  traversePtrs f expr = case expr of
+    App e1 e2 -> App  <$> tp f e1 <*> tp f e2
+    Atom x    -> Atom <$> tp f x
+    Op  e     -> Op   <$> traverse (tp f) e
+    Hof e     -> Hof  <$> traverse (tp f) e
+    Case e alts resultTy ->
+      Case <$> tp f e <*> traverse (tp f) alts <*> tp f resultTy
+
+instance (HasPtrs a, HasPtrs b) => HasPtrs (Abs a b) where
+  traversePtrs f (Abs b body) = Abs <$> tp f b <*> tp f body
+
+instance HasPtrs Block where
+  traversePtrs f (Block decls result) =
+    Block <$> traverse (tp f) decls <*> tp f result
+
+instance HasPtrs Decl where
+  traversePtrs f (Let ann b body) = Let ann <$> tp f b <*> tp f body
+
+instance (HasPtrs a, HasPtrs b) => HasPtrs (a, b) where
+  traversePtrs f (x, y) = (,) <$> tp f x <*> tp f y
+
+instance HasPtrs eff => HasPtrs (ArrowP eff) where
+  traversePtrs f arrow = case arrow of
+    PlainArrow eff -> PlainArrow <$> tp f eff
+    _ -> pure arrow
+
+instance (HasPtrs a, HasPtrs b) => HasPtrs (ExtLabeledItems a b) where
+  traversePtrs f (Ext items t) =
+    Ext <$> traverse (tp f) items <*> traverse (tp f) t
+
+instance HasPtrs DataConRefBinding where
+  traversePtrs f (DataConRefBinding b ref) =
+    DataConRefBinding <$> tp f b <*> tp f ref
+
+instance HasPtrs Atom where
+  traversePtrs f atom = case atom of
+    Var v   -> Var <$> traverse (tp f) v
+    Lam lam -> Lam <$> tp f lam
+    Pi  ty  -> Pi  <$> tp f ty
+    TC  tc  -> TC  <$> traverse (tp f) tc
+    Con (Lit (PtrLit ptrTy ptr)) -> (Con . Lit . PtrLit ptrTy) <$> f ptrTy ptr
+    Con con -> Con <$> traverse (tp f) con
```

Comment:
You probably don't want to traverse `SumAsCon` constructors, or else we'll end up with the usual bug of dereferencing garbage pointers.

Reply:
Good catch. Just pushed 88c2648 which initializes pointer memory to zero so that we save uninitialized pointers as `NullPtr`.